### PR TITLE
Option to disable mipmaps on rendertargets.  2 additional viewport usage modes.

### PR DESCRIPTION
**Pull request by jordo**
This PR is a result of performance profiling which shows a significant performance impact when sampling `SCREEN_TEXTURE` especially on mobile devices (profiling on iPhone 11).  

Using the `SCREEN_TEXTURE` ended up causing a mipmap generation pass, which runs a 2-pass gaussian blur filter for every generated mipmap level.  This ended up being a significant performance impact both on the cpu side of this loop, and gpu.  Especially on mobile.

Considering, most times when a developer samples the `SCREEN_TEXTURE`, you are doing so in screen UV coordinates (`texture(SCREEN_TEXTURE, SCREEN_UV)`), it is really a lot of additional work to generate the mipmap levels for not much benefit.

The PR adds two additional usage modes to viewport, which then set a rendertarget flag which will omit mipmap generation for that render target.  This is useful globally for the main framebuffer allocation via the project setting `rendering/quality/intended_usage/framebuffer_allocation`, but is also exposed per viewport, so you can adjust on a per viewport basis if desired.  The additional enums extend the existing viewport usage modes, so they shouldn't conflict with existing project settings.

Screen Shot 2021-09-30 at 5 03 54 PM

This resulted in an approximately 25% improvement in CPU utilization on a few test mobile devices as well as saved ~1-2ms gpu time on these devices as well.  

Please note, I am unfamiliar with how master storage layers are managed, but the concepts should apply to master as well.  (It should be possible to disable mipmap generation for render targets, and also be able to configure this as a potential viewport usage mode).

### CODE REVIEW

File: drivers/gles2/rasterizer_storage_gles2.cpp [hunk 1]
```diff
@@ -5126,7 +5126,7 @@ void RasterizerStorageGLES2::_render_target_allocate(RenderTarget *rt) {
 	}
 
 	// Allocate mipmap chains for post_process effects
-	if (!rt->flags[RasterizerStorage::RENDER_TARGET_NO_3D] && rt->width >= 2 && rt->height >= 2) {
+	if (!rt->flags[RasterizerStorage::RENDER_TARGET_NO_3D] && !rt->flags[RasterizerStorage::RENDER_TARGET_NO_MIPMAPS] && rt->width >= 2 && rt->height >= 2) {
```
Comment (clayjohn) [inline drivers/gles2/rasterizer_storage_gles2.cpp:5129]: These are the mipmaps for post processing effects like glow. I don't think they should be disabled when "no mipmaps" is selected.

File: drivers/gles3/rasterizer_storage_gles3.cpp [hunk 2]
```diff
@@ -7171,85 +7171,140 @@ void RasterizerStorageGLES3::_render_target_allocate(RenderTarget *rt) {
 	}
 
 	if (!rt->flags[RENDER_TARGET_NO_SAMPLING] && rt->width >= 2 && rt->height >= 2) {
-		for (int i = 0; i < 2; i++) {
-			ERR_FAIL_COND(rt->effects.mip_maps[i].sizes.size());
+		if (rt->flags[RENDER_TARGET_NO_MIPMAPS]) {
```
Comment (clayjohn) [inline drivers/gles3/rasterizer_storage_gles3.cpp:7174]: I think these mip maps are also reused by glow

**Review by clayjohn — COMMENTED** [review]
I don't think that disabling the ``SCREEN_TEXTURE`` mipmaps should disable 3D effects as well. I think that it would be rather surprising for a user to lose post-processing effects when disabling mipmaps. 

That being said, I expect that some users would like to disable both ``SCREEN_TEXTURE`` mipmaps and post processing effects. 

In my opinion the best approach would be to just store a single flag called ``generate_mipmaps" or something similar that can be toggled on and off.

File: drivers/gles2/rasterizer_storage_gles2.cpp (hunk 1 above)
Comment (jordo) [reply drivers/gles2/rasterizer_storage_gles2.cpp:5129]: Yep that's what I intended.  Post process effects should still run, just not with mipmap sampler.  

See https://github.com/godotengine/godot/pull/53275/files#diff-e61a2716371ede6bb16219d6b8f676a851fdeb08b948798cf11a4a0041eb6808R681-R688

Gles2 I might have to take another look at....  As it looks a little different on second look.  But that's what currently happening on gles3.

> In my opinion the best approach would be to just store a single flag called ``generate_mipmaps" or something similar that can be toggled on and off.

RENDER_TARGET_NO_MIPMAPS is basically this flag...  Or you would want another separate flag per viewport?  Either way, there would need to be a flag on RenderTarget, but it doesn't matter to me how it's exposed externally in the viewport.  It looked like viewport usage_mode is design to moderate the RenderTarget flag permutations here:

https://github.com/godotengine/godot/pull/53275/files#diff-e61a2716371ede6bb16219d6b8f676a851fdeb08b948798cf11a4a0041eb6808R657-R698

Comment (jordo) [reply drivers/gles2/rasterizer_storage_gles2.cpp:5129]: Ya I can confirm glow is broken without mipmaps.  The main issue is, when keeping locality to just the SCREEN_TEXTURE issue, you still have to end up setting up the screen's RenderTarget correctly, which ur right affects glow.  hmmmm

**Comment by jordo** [comment]
So ya, I guess the issue is changing the Screen texture's RenderTarget (or any RenderTarget) for that matter and disabling mipmaps, is going to disable 3d effects as well.  I don't think there is any way around this.  If you setup a RenderTarget without mipmaps, 3d effects work as that's a requirement

**Comment by jordo** [comment]
Just omitting (not generating) the mipmaps, for the screen's RenderTarget (who's storage has mipmaps enabled,) leads to errors like this and unexpected artifacts: 

Screen Shot 2021-09-30 at 9 10 42 PM

So we can't just flag this out at the screen level.  It has to setup the correct rendertarget.  

So I guess I'm back to thinking if you disable mipmaps on a render target (screen or other), you will end up disabling 3d effects as well.

**Comment by jordo** [comment]
closed, going to re-tweak for SCREEN setting only.

**Comment by Calinou** [comment]
> closed, going to re-tweak for SCREEN setting only.

Did you manage to do this?

**Pull request closed without merging**